File: TB2S2016_CSharp/TB2S2016_CSharp/Conta.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace TB2S2016_CSharp
{
    internal class Conta
    {
        private string agencia;
        private String numero;
        private string nome;
        private DateTime dataAbertura;
        private double saldo;
        private Double limite;

        public String Numero
        {
            get { return this.numero; }
            set { this.numero = value.Replace("X", "0"); }
        }


        public string Nome
        {
            get { return nome; }
            set { nome = value; }
        }

        public double Saldo
        {
            get
            {
                return saldo;
            }

            set
            {
                saldo = value;
            }
        }

        public DateTime DataAbertura
        {
            get
            {
                return dataAbertura;
            }

            set
            {
                dataAbertura = value;
            }
        }

        public double Limite
        {
            get
            {
                return limite;
            }

            set
            {
                limite = value;
            }
        }

        public string Agencia
        {
            get
            {
                return agencia;
            }

            set
            {
                agencia = value;
            }
        }

        public Conta()
        {
            this.limite = 100;
        }

        public Conta(string agencia, string numero)
            : this()
        {
            this.Agencia = agencia;
            this.Numero = numero;
        }

        public Conta(string agencia, string numero, string nome)
            : this(agencia, numero)
        {
            this.Nome = nome;
        }

        public Boolean Deposita(double valor)
        {
            if (valor <= 0.0)
                return false;

            this.Saldo += valor;
            return true;
        }

        public bool Saca(double valor)
        {
            if (this.Saldo < valor)
                return false;

            this.Saldo -= valor;
            return true;
        }

        public bool Transfere(double valor, Conta conta)
        {
            if (this.Saca(valor))
            {
                return conta.Deposita(valor);
            }
            return false;
        }
    }
}

File: TB2S2016_Java/src/tb2s2016_java/Conta.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package tb2s2016_java;

import java.util.Date;

/**
 *
 * @author antonio.lopes
 */
public class Conta {
 
    private String agencia;
    private String numero;
    private String nome;
    private Date dataAbertura;
    private double saldo;
    private double limite;
    
    public String getAgencia() {
        return this.agencia;
    }
    public void setAgencia(String agencia) {
        this.agencia = agencia;
    }   

    public String getNumero() {
        return numero;
    }

    public void setNumero(String numero) {
        this.numero = numero;
    }

    public Date getDataAbertura() {
        return dataAbertura;
    }

    public void setDataAbertura(Date dataAbertura) {
        this.dataAbertura = dataAbertura;
    }

    public double getSaldo() {
        return saldo;
    }

    public void setSaldo(double saldo) {
        this.saldo = saldo;
    }

    public double getLimite() {
        return limite;
    }

    public void setLimite(double limite) {
        this.limite = limite;
    }
    
    public String getNome() {
        return nome;
    }
    public void setNome(String nome) {
        this.nome = nome.toUpperCase();
    }
        
    Conta() {
        this.setLimite(100.0);
    }
    
    public Conta(String agencia, String numero) {
        this();
        this.setAgencia(agencia);
        this.numero = numero;
    }
    
    public Conta(String agencia, String numero, String nome) {
        this(agencia, numero);
        this.setNome(nome);
    }
    
    public Conta(String agencia, String numero,
            String nome, double limite) {
        this(agencia, numero, nome);
        this.setLimite(limite);
    }
    
    boolean deposita(double valor)
        throws ContaException {
        if (valor <= 0.0)
            throw new ContaException("Valor do depósito inválido!");
        
        this.saldo += valor;
        return true;
    }
    
    boolean saca(double valor)
            throws ContaException
    {
        if (this.saldo < valor)
            throw new ContaException("Sem saldo!");
        
        this.saldo -= valor;
        return true;
    }
    
    boolean transferePara(Conta contadestino, double valor)
            throws ContaException
    {
        if (!saca(valor))
            return false;
        
        return contadestino.deposita(valor);
    }
}
